ui/app/components/app/modals/external-sign-modal/external-sign-signer-airsign.js

```diff
@@ -0,0 +1,142 @@
+
+  /*
+   * Returns the number of QR codes that have to be transmitted.
+   * @returns {Number} the number of QR codes to be transmitted
+   */
+  getQrCount = () => {
+    return 1
+  }
+
+  /**
+   * Extracts the necessary URI from the signature provided by external signer.
+   * @param {String} signatureUri The string read by QR code scanner from external signer
+   * @return {String}
+   * The function must throw an exception if the signatureUri is invalid.
+   */
+  decodeSignatureString = (signatureUri) => {
+    if (/^eths:\/\?s=0x[0-9a-f]{130}$/i.test(signatureUri)) {
+      return signatureUri.replace(/^eths:\/\?s=/i, '')
+    } else {
+      throw new Error('Invalid signature')
+    }
+  }
+
+  /**
+   * Extracts signature from signer string.
+   * @param {String} signedUri
+   * @return {Object} {r, s, v} where r, s, and v are the signature elements
+   *
+   * This function must be able to process the return value of decodeSignatureString() function.
+   */
+  decodeSigned = (signedUri) => {
+    signedUri = ethUtil.stripHexPrefix(signedUri)
+    const r = ethUtil.toBuffer('0x' + signedUri.substr(0, 64))
+    const s = ethUtil.toBuffer('0x' + signedUri.substr(64, 64))
+    const v = ethUtil.toBuffer('0x' + signedUri.substr(128, 2))
+    return {r: r, s: s, v: v}
+  }
+
+
+  /**
+   * Encode the to be signed tx or message in a format readable by external signer.
+   * @param {Number} qrNum the number of QR to be displayed starting from 0
+   * @return {String} the encoded tx or message ready for being encoded in QR
+   */
+  encodeSignRequestUri = (qrNum) => {
+    if (qrNum !== 0) {
+      throw new Error('Multiple QR code transmission not supported yet.')
+    }
+    const fromAddress = ethUtil.stripHexPrefix(this.signable.from)
+    const partFrom = fromAddress.substr(0, 4) +
+      fromAddress.substr(18, 2) + fromAddress.slice(-4)
+    switch (this.signable.type) {
+      case 'sign_message':
+        return 'eths:/?m=' + partFrom + this.signable.payload
+      case 'sign_personal_message':
+        return 'eths:/?p=' + partFrom + this.signable.payload
+      case 'sign_typed_data':
+        return 'eths:/?y=' + partFrom + this.signable.payload
+      case 'sign_transaction':
+        const tx = this.signable.payload
+        var chainId = tx.chainId
+        var chainCode
+        switch (chainId) {
+          case 1:
+            chainCode = 0
+            break
+          case 61:
+            chainCode = 1
+            break
+          case 2:
+            chainCode = 3 // Morden testnet
+            break
+          case 3:
+            chainCode = 4
+            break
+          case 4:
+            chainCode = 5
+            break
+          case 42:
+            chainCode = 6
+            break
+          case 77:
+            chainCode = 7
+            break
+          case 99:
+            chainCode = 9
+            break
+          case 7762959:
+            chainCode = 10
+            break
+          default:
```

Comment:
where did you get all these numbers? we should put a reference

Reply:
Reference is there. Done.